internal/controlplane/handlers_entities.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2023 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+
+	"github.com/stacklok/minder/internal/db"
+	"github.com/stacklok/minder/internal/engine"
+	"github.com/stacklok/minder/internal/events"
+	"github.com/stacklok/minder/internal/logger"
+	"github.com/stacklok/minder/internal/providers"
+	"github.com/stacklok/minder/internal/reconcilers/messages"
+	"github.com/stacklok/minder/internal/util"
+	pb "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// ReconcileEntityRegistration reconciles the registration of an entity.
+//
+// Currently, this method only supports repositories but is intended to be
+// generic and handle all types of entities.
+// Todo: Utilise for other entities when such are supported.
+func (s *Server) ReconcileEntityRegistration(
+	ctx context.Context,
+	in *pb.ReconcileEntityRegistrationRequest,
+) (*pb.ReconcileEntityRegistrationResponse, error) {
+	l := zerolog.Ctx(ctx).With().Logger()
+
+	entityCtx := engine.EntityFromContext(ctx)
+	projectID := entityCtx.Project.ID
+
+	logger.BusinessRecord(ctx).Project = projectID
+
+	providerName := in.GetContext().GetProvider()
+	provs, errorProvs, err := s.providerManager.BulkInstantiateByTrait(ctx, projectID, db.ProviderTypeRepoLister, providerName)
+	if err != nil {
+		pErr := providers.ErrProviderNotFoundBy{}
+		if errors.As(err, &pErr) {
+			return nil, util.UserVisibleError(codes.NotFound, "no suitable provider found, please enroll a provider")
+		}
+		return nil, providerError(err)
+	}
+
+	for providerName, provider := range provs {
+		// Explicitly fetch the provider here as we need its ID for posting the event.
+		pvr, err := s.providerStore.GetByName(ctx, projectID, providerName)
+		if err != nil {
+			errorProvs = append(errorProvs, providerName)
+			continue
+		}
+
+		repos, err := s.fetchRepositoriesForProvider(ctx, projectID, providerName, provider)
+		if err != nil {
+			errorProvs = append(errorProvs, providerName)
```

Comment:
since the handler only errors out if all providers fail could we `zerolog` the error message? (I guess there would realistically be only one provider though)

Reply:
Line 75 already ships the error with the log as a structured field, do you mean something more detailed than that?